File: python/src/wdl_model/model/processors/wdl_processor.py
"""Visitor-style processor contract for the full Python WDL object model."""

from __future__ import annotations

from abc import ABC, abstractmethod

from wdl_model.model.definitions import (
    WdlEnum,
    WdlStruct,
    WdlStructMember,
    WdlTask,
    WdlWorkflow,
)
from wdl_model.model.sections import (
    WdlCommand,
    WdlInput,
    WdlMetadata,
    WdlOutput,
    WdlParameterMetadata,
    WdlRequirements,
    WdlRuntime,
    WdlTaskHints,
    WdlWorkflowHints,
)
from wdl_model.model.statements import (
    WdlBoundDeclaration,
    WdlCall,
    WdlConditional,
    WdlImportMembers,
    WdlImportStandard,
    WdlImportStar,
    WdlScatter,
)

from ..wdl_document import WdlDocument
from ..wdl_version import WdlVersion


class WdlProcessor(ABC):
    """Abstract processor contract for documents, definitions, sections, and statements.

    Most callers should subclass ``WdlProcessorBase`` rather than implement this interface from
    scratch.
    """

    @abstractmethod
    def processDocument(self, node: WdlDocument) -> None: ...

    @abstractmethod
    def processVersion(self, ctx: WdlDocument, node: WdlVersion) -> None: ...

    @abstractmethod
    def processImportStandard(
        self, ctx: WdlDocument, node: WdlImportStandard
    ) -> None: ...

    @abstractmethod
    def processImportMembers(
        self, ctx: WdlDocument, node: WdlImportMembers
    ) -> None: ...

    @abstractmethod
    def processImportStar(self, ctx: WdlDocument, node: WdlImportStar) -> None: ...

    @abstractmethod
    def processEnum(self, ctx: WdlDocument, node: WdlEnum) -> None: ...

    @abstractmethod
    def processStruct(self, ctx: WdlDocument, node: WdlStruct) -> None: ...

    @abstractmethod
    def processStructMember(self, ctx: WdlStruct, node: WdlStructMember) -> None: ...

    @abstractmethod
    def processStructParameterMetadata(
        self, ctx: WdlStruct, node: WdlParameterMetadata
    ) -> None: ...

    @abstractmethod
    def processStructMetadata(self, ctx: WdlStruct, node: WdlMetadata) -> None: ...

    @abstractmethod
    def processTask(self, ctx: WdlDocument, node: WdlTask) -> None: ...

    @abstractmethod
    def processTaskDeclaration(
        self, ctx: WdlTask, node: WdlBoundDeclaration
    ) -> None: ...

    @abstractmethod
    def processTaskInput(self, ctx: WdlTask, node: WdlInput) -> None: ...

    @abstractmethod
    def processTaskOutput(self, ctx: WdlTask, node: WdlOutput) -> None: ...

    @abstractmethod
    def processTaskCommand(self, ctx: WdlTask, node: WdlCommand) -> None: ...

    @abstractmethod
    def processTaskParameterMetadata(
        self, ctx: WdlTask, node: WdlParameterMetadata
    ) -> None: ...

    @abstractmethod
    def processTaskMetadata(self, ctx: WdlTask, node: WdlMetadata) -> None: ...

    @abstractmethod
    def processTaskRequirements(self, ctx: WdlTask, node: WdlRequirements) -> None: ...

    @abstractmethod
    def processTaskRuntime(self, ctx: WdlTask, node: WdlRuntime) -> None: ...

    @abstractmethod
    def processTaskHints(self, ctx: WdlTask, node: WdlTaskHints) -> None: ...

    @abstractmethod
    def processWorkflow(self, ctx: WdlDocument, node: WdlWorkflow) -> None: ...

    @abstractmethod
    def processWorkflowDeclaration(
        self, ctx: WdlWorkflow, node: WdlBoundDeclaration
    ) -> None: ...

    @abstractmethod
    def processWorkflowInput(self, ctx: WdlWorkflow, node: WdlInput) -> None: ...

    @abstractmethod
    def processWorkflowOutput(self, ctx: WdlWorkflow, node: WdlOutput) -> None: ...

    @abstractmethod
    def processWorkflowMetadata(self, ctx: WdlWorkflow, node: WdlMetadata) -> None: ...

    @abstractmethod
    def processWorkflowParameterMetadata(
        self, ctx: WdlWorkflow, node: WdlParameterMetadata
    ) -> None: ...

    @abstractmethod
    def processWorkflowCall(self, ctx: WdlWorkflow, node: WdlCall) -> None: ...

    @abstractmethod
    def processWorkflowConditional(
        self, ctx: WdlWorkflow, node: WdlConditional
    ) -> None: ...

    @abstractmethod
    def processWorkflowScatter(self, ctx: WdlWorkflow, node: WdlScatter) -> None: ...

    @abstractmethod
    def processWorkflowHints(
        self, ctx: WdlWorkflow, node: WdlWorkflowHints
    ) -> None: ...

File: java/src/main/java/com/myriad/wdl/model/processors/WdlProcessor.java
package com.myriad.wdl.model.processors;

import com.myriad.wdl.model.WdlDocument;
import com.myriad.wdl.model.WdlVersion;
import com.myriad.wdl.model.definitions.WdlEnum;
import com.myriad.wdl.model.definitions.WdlStruct;
import com.myriad.wdl.model.definitions.WdlStruct.WdlStructMember;
import com.myriad.wdl.model.definitions.WdlTask;
import com.myriad.wdl.model.definitions.WdlWorkflow;
import com.myriad.wdl.model.sections.WdlCommand;
import com.myriad.wdl.model.sections.WdlHints.WdlTaskHints;
import com.myriad.wdl.model.sections.WdlHints.WdlWorkflowHints;
import com.myriad.wdl.model.sections.WdlInput;
import com.myriad.wdl.model.sections.WdlMetadataBase.WdlMetadata;
import com.myriad.wdl.model.sections.WdlMetadataBase.WdlParameterMetadata;
import com.myriad.wdl.model.sections.WdlOutput;
import com.myriad.wdl.model.sections.WdlRequirements;
import com.myriad.wdl.model.sections.WdlRuntime;
import com.myriad.wdl.model.statements.WdlCall;
import com.myriad.wdl.model.statements.WdlConditional;
import com.myriad.wdl.model.statements.WdlDeclaration.WdlBoundDeclaration;
import com.myriad.wdl.model.statements.WdlImport.WdlImportMembers;
import com.myriad.wdl.model.statements.WdlImport.WdlImportStandard;
import com.myriad.wdl.model.statements.WdlImport.WdlImportStar;
import com.myriad.wdl.model.statements.WdlScatter;

/**
 * Visitor-style processor contract for the full Java WDL object model.
 *
 * <p>Implement this interface when you want explicit callbacks for the major document, definition,
 * section, and workflow statement nodes. {@link WdlProcessorBase} provides the usual source-order
 * traversal so most callers should extend that instead of implementing everything directly.
 */
public interface WdlProcessor {

  void processDocument(WdlDocument node);

  void processVersion(WdlDocument ctx, WdlVersion node);

  void processImport(WdlDocument ctx, WdlImportStandard node);

  void processImport(WdlDocument ctx, WdlImportMembers node);

  void processImport(WdlDocument ctx, WdlImportStar node);

  void processEnum(WdlDocument ctx, WdlEnum node);

  void processStruct(WdlDocument ctx, WdlStruct node);

  void processStructMember(WdlStruct ctx, WdlStructMember node);

  void processStructParameterMetadata(WdlStruct ctx, WdlParameterMetadata node);

  void processStructMetadata(WdlStruct ctx, WdlMetadata node);

  void processTask(WdlDocument ctx, WdlTask node);

  void processTaskDeclaration(WdlTask ctx, WdlBoundDeclaration node);

  void processTaskInput(WdlTask ctx, WdlInput node);

  void processTaskOutput(WdlTask ctx, WdlOutput node);

  void processTaskCommand(WdlTask ctx, WdlCommand node);

  void processTaskParameterMetadata(WdlTask ctx, WdlParameterMetadata node);

  void processTaskMetadata(WdlTask ctx, WdlMetadata node);

  void processTaskRequirements(WdlTask ctx, WdlRequirements node);

  void processTaskRuntime(WdlTask ctx, WdlRuntime node);

  void processTaskHints(WdlTask ctx, WdlTaskHints node);

  void processWorkflow(WdlDocument ctx, WdlWorkflow node);

  void processWorkflowDeclaration(WdlWorkflow ctx, WdlBoundDeclaration node);

  void processWorkflowInput(WdlWorkflow ctx, WdlInput node);

  void processWorkflowOutput(WdlWorkflow ctx, WdlOutput node);

  void processWorkflowMetadata(WdlWorkflow ctx, WdlMetadata node);

  void processWorkflowParameterMetadata(WdlWorkflow ctx, WdlParameterMetadata node);

  void processWorkflowCall(WdlWorkflow ctx, WdlCall node);

  void processWorkflowConditional(WdlWorkflow ctx, WdlConditional node);

  void processWorkflowScatter(WdlWorkflow ctx, WdlScatter node);

  void processWorkflowHints(WdlWorkflow ctx, WdlWorkflowHints node);
}
